pkg/config/policyparser.go

```diff
@@ -0,0 +1,347 @@
+/*
+Copyright IBM Corp. 2017 All Rights Reserved.
+
+SPDX-License-Identifier: Apache-2.0
+*/
+
+package config
+
+import (
+	"fmt"
+	"reflect"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/Knetic/govaluate"
+	"github.com/gogo/protobuf/proto"
+	cb "github.com/hyperledger/fabric-protos-go/common"
+	mb "github.com/hyperledger/fabric-protos-go/msp"
+)
+
+// Gate values
+const (
+	GateAnd   = "And"
+	GateOr    = "Or"
+	GateOutOf = "OutOf"
+)
+
+// Role values for principals
+const (
+	RoleAdmin   = "admin"
+	RoleMember  = "member"
+	RoleClient  = "client"
+	RolePeer    = "peer"
+	RoleOrderer = "orderer"
+)
+
+var (
+	regex = regexp.MustCompile(
+		fmt.Sprintf("^([[:alnum:].-]+)([.])(%s|%s|%s|%s|%s)$",
+			RoleAdmin, RoleMember, RoleClient, RolePeer, RoleOrderer),
+	)
+	regexErr = regexp.MustCompile("^No parameter '([^']+)' found[.]$")
+)
+
+// outof is a stub function - it returns the same string as it's passed
+// This will be evaluated by second/third passes to convert to a proto policy
+func outof(args ...interface{}) (interface{}, error) {
+	toret := "outof("
+	if len(args) < 2 {
+		return nil, fmt.Errorf("expected at least two arguments to NOutOf, but got %d", len(args))
+	}
+
+	arg0 := args[0]
+	// govaluate treats all numbers as float64 only. But and/or may pass int/string. Allowing int/string for flexibility of caller
+	if n, ok := arg0.(float64); ok {
+		toret += strconv.Itoa(int(n))
+	} else if n, ok := arg0.(int); ok {
+		toret += strconv.Itoa(n)
+	} else if n, ok := arg0.(string); ok {
+		toret += n
+	} else {
+		return nil, fmt.Errorf("unexpected type %s", reflect.TypeOf(arg0))
+	}
+
+	for _, arg := range args[1:] {
+		toret += ", "
+		switch t := arg.(type) {
+		case string:
+			if regex.MatchString(t) {
+				toret += "'" + t + "'"
+			} else {
+				toret += t
+			}
+		default:
+			return nil, fmt.Errorf("unexpected type %s", reflect.TypeOf(arg))
+		}
+	}
+	return toret + ")", nil
+}
+
+func and(args ...interface{}) (interface{}, error) {
+	args = append([]interface{}{len(args)}, args...)
+	return outof(args...)
+}
+
+func or(args ...interface{}) (interface{}, error) {
+	args = append([]interface{}{1}, args...)
+	return outof(args...)
+}
+
+func firstPass(args ...interface{}) (interface{}, error) {
+	toret := "outof(ID"
+	for _, arg := range args {
+		toret += ", "
+		switch t := arg.(type) {
+		case string:
+			if regex.MatchString(t) {
+				toret += "'" + t + "'"
+			} else {
+				toret += t
+			}
+		case float32:
+		case float64:
+			toret += strconv.Itoa(int(t))
+		default:
+			return nil, fmt.Errorf("unexpected type %s", reflect.TypeOf(arg))
+		}
+	}
+
+	return toret + ")", nil
+}
+
+func secondPass(args ...interface{}) (interface{}, error) {
+	/* general sanity check, we expect at least 3 args */
```

Comment:
super nit (and I know it's likely code that was copied as-is): feels strange to see /* */ block comments in just this one file (and actually just this one function) where everywhere else we use basic // line comments.

Reply:
fixed 🔧